### PR TITLE
Ensure object-fit, object-position, and layout=fill are set on cover block image/video background; omit playsinline attribute in video sanitizer

### LINKED ISSUE #5914: Cover block issue with Wordpress 5.7
## Bug Description

The 5.7 cover block now uses an img tag instead of background image. This causes problems with  amp  object-fit: contain, image does not fill background.

## Expected Behaviour

Image should cover entire background. I think object-fit: cover would be better.for the cover block

## Steps to reproduce

Create a cover block in wordpress 5.7 with a background image. You can check out an example here.  https://www.ryansmithphotography.com/myrtle-beach-real-estate-photography/

## Additional context

 - WordPress version: 5.7 RC1
 - Plugin version:
 - Gutenberg plugin version (if applicable):
 - AMP plugin template mode:
 - PHP version:
 - OS:
 - Browser: [e.g. chrome, safari]

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

Fixes #5914

* Fix up the _first_ background `img`/`video` element in a Cover block by adding these attributes (which are passed through to the underlying `amp-img`/`amp-video`):
  * `object-fit=cover` (by definition)
  * `layout=fill` (to avoid the need to look up dimensions)
  * `object-position` (when a focal point has been selected)
* Update the Video sanitizer to omit the `playsinline`/`webkit-playsinline` attributes since they are not valid AMP. The `amp-video` extension adds these automatically to the underlying `video` element at runtime. This mirrors the behavior for when `loading=lazy` is added to an `img`: the image sanitizer just omits it because it is the default behavior in AMP.
* Allow a `$message` parameter to be passed to the `MarkupComparison` assertion methods.

## Checklist

- [x] Add tests.
- [x] Test with oldest version of Gutenberg that had focal point.
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/embeds/class-amp-core-block-handler.php [hunk 1]
```diff
@@ -215,23 +215,47 @@ public function ampify_video_block( $block_content, $block ) {
...
-		if ( isset( $block['attrs']['backgroundType'] ) && 'video' === $block['attrs']['backgroundType'] ) {
-			$block_content = preg_replace(
-				'/(?<=<video\s)/',
-				'layout="fill" object-fit="cover" ',
-				$block_content
-			);
-		}
-		return $block_content;
+		return preg_replace_callback(
+			'/<(?:img|video)(?= )[^>]* class="[^"]*?wp-block-cover__(?:image|video)-background[^"]*?"[^>]*(?=>)/',
```
Comment (westonruter) [inline includes/embeds/class-amp-core-block-handler.php]:
```suggestion
			'/<(?:img|video)(?= )[^>]*? class="[^"]*?wp-block-cover__(?:image|video)-background[^"]*?"[^>]*(?=>)/',
```

File: includes/embeds/class-amp-core-block-handler.php [hunk 2]
```diff
@@ -224,14 +226,36 @@ public function ampify_video_block( $block_content, $block ) {
...
-		if ( isset( $block['attrs']['backgroundType'] ) && 'video' === $block['attrs']['backgroundType'] ) {
-			$block_content = preg_replace(
-				'/(?<=<video\s)/',
-				'layout="fill" object-fit="cover" ',
-				$block_content
-			);
-		}
-		return $block_content;
+		return preg_replace_callback(
+			'#<(?:img|video)(?= )[^>]*? class="[^"]*?wp-block-cover__(?:image|video)-background[^"]*?"#',
```
Comment (pierlon) [inline includes/embeds/class-amp-core-block-handler.php]: This regex would match class names like `my-custom-class-wp-block-cover__image-background` or `wp-block-cover__video-background-plugin-baz`, which is unintentional. Here's one way to resolve the issue:

```suggestion
			'#<(?:img|video)(?= )[^>]*? class=(?:"|"[^"]*? )wp-block-cover__(?:image|video)-background(?:"| [^"]*?")#',
```

File: includes/embeds/class-amp-core-block-handler.php [hunk 3]
```diff
@@ -224,14 +226,36 @@ public function ampify_video_block( $block_content, $block ) {
...
+			static function ( $matches ) use ( $block ) {
+				$replacement = $matches[0];
+
+				// The background image/video for the cover block by definition needs object-fit="cover" on the resulting amp-ing/amp-video.
+				$replacement .= ' object-fit="cover"';
+
+				// Add the fill layout to skip the needlessly obtaining the dimensions.
```
Comment (pierlon) [inline includes/embeds/class-amp-core-block-handler.php]:
```suggestion
				// Add the fill layout to skip needlessly obtaining the dimensions.
```

File: includes/embeds/class-amp-core-block-handler.php [hunk 4]
```diff
@@ -224,14 +226,36 @@ public function ampify_video_block( $block_content, $block ) {
...
+				$replacement .= ' layout="fill"';
+
+				// Add object-position from the block attribute's to add to the img/video to be copied onto the amp-img/amp-video.
```
Comment (pierlon) [inline includes/embeds/class-amp-core-block-handler.php]:
```suggestion
				// Add object-position from the block's attributes to add to the img/video to be copied onto the amp-img/amp-video.
```

**Review by pierlon — CHANGES_REQUESTED** [review]

File: includes/embeds/class-amp-core-block-handler.php (hunk 2 above)
Comment (westonruter) [reply includes/embeds/class-amp-core-block-handler.php]: This seems to be a bit better:

```suggestion
			'#<(?:img|video)(?= )[^>]*? class="(?:[^"]*? )?wp-block-cover__(?:image|video)-background(?: [^"]*?)?"#',
```

(Factoring out the common double quote marks.)

Comment (pierlon) [reply includes/embeds/class-amp-core-block-handler.php]: That works, and takes less steps to match too!

**Review by pierlon — APPROVED** [review]

**Comment by westonruter** [comment]
@pierlon There was one last case not yet accounted for: when a cover block doesn't use an `img` element at all (when the background is fixed or repeated). When a cover block was nested inside another cover block, the result was duplication of attributes on the nested cover block. This is addressed in 532c7732663c98b09ca9fe9d1f1fec166a5a4225.

File: tests/php/test-class-amp-core-block-handler.php [hunk 5]
```diff
@@ -178,12 +178,16 @@ public function get_data_for_test_ampify_cover_block() {
...
...-content/uploads/2021/02/stamp_video_dog.mp4\"></video><div class=\"wp-block-cover__inner-container\">\n<p class=\"has-text-align-center has-large-font-size\">Nested Image and Video</p>\n\n\n\n<figure class=\"wp-block-image size-large\"><img src=\"https://example.com/wp-content/uploads/2021/02/image.MP_-768x1024.jpg\" alt=\"\" class=\"wp-image-48\"/></figure>\n\n\n\n<figure class=\"wp-block-video\"><video controls src=\"https://example.com/wp-content/uploads/2021/02/stamp_video_dog.mp4\"></video></figure>\n</div></div>\n\n\n\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(https://example.com/wp-content/uploads/2021/02/image.MP_-scaled.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p class=\"has-text-align-center has-large-font-size\">Nested Cover</p>\n\n\n\n<div class=\"wp-block-cover has-background-dim\"><video class=\"wp-block-cover__video-background\" object-fit=\"cover\" layout=\"fill\" autoplay muted loop src=\"https://example.com/wp-content/uploads/2021/02/stamp_video_dog.mp4\"></video><div class=\"wp-block-cover__inner-container\">\n<p class=\"has-text-align-center has-large-font-size\">Inner video!</p>\n</div></div>\n</div></div>\n",
```
Comment (westonruter) [inline tests/php/test-class-amp-core-block-handler.php:181]: Notice the fix here:

```diff
- object-fit=\"cover\" layout=\"fill\" object-fit=\"cover\" layout=\"fill\"
+ object-fit=\"cover\" layout=\"fill\"
```

**Pull request merged**